R/as-tibble.R

```diff
@@ -1,26 +1,77 @@
+#' Flatten a Crunch Cube
+#'
+#' Crunch Cubes can be expressed as a long data frame instead of a
+#' multidimensional array. In this form each dimension of the cube is a variable
+#' and the cube values are expressed columns for each measure. This is useful
+#' both to better understand what each entry of a cube represents, and to work
+#' with the cube result using tidyverse tools.
+#'
+#' @param x a CrunchCube
+#' @param ... futher arguments passed on to `tibble::as_tibble()`
+#'
 #' @export
 #' @importFrom tibble as_tibble
+#' @importFrom crunch getDimTypes
+#' @importFrom dplyr bind_cols
+#' @importFrom purrr map2 map reduce
+#' @importFrom stringr str_extract
 as_tibble.CrunchCube <- function (x, ...) {
     ## TODO: Consider using `dplyr::tbl_cube` class
-    ## TODO: NA behavior? Add way to get missings not missing. (This uses
-    ## as.array to remove extraneous dims, and doing so applies the '@useNA'
-    ## from the cube)
-    ## TODO: deal with naming for array (and NUMR) variable dims (var.categories?)
+    
+    dnames <- dimnames(x@arrays$.unweighted_counts)
+    measures <- names(x@arrays)
 
-    ## 1) get the measures
-    ## TODO: better access methods for secondary measures in Cube
-    ## TODO: if weighted query, include ".unweighted_counts"
-    measure_names <- setdiff(names(x@arrays), ".unweighted_counts")
-    vals <- sapply(measure_names, function (m) {
-        as.vector(crunch:::cubeToArray(x, m))
-    }, simplify=FALSE)
+    # Cubes can have multiple measures, which are represented as their own column
+    # in the tibble.
+    measure_vals <- sapply(measures, function(y) {
+        as.vector(x@arrays[[y]])
+    }, simplify = FALSE)
+    
+    names(measure_vals)[names(measure_vals) == ".unweighted_counts"] <- "row_count"
 
-    ## 2) If there are dimnames, expand.grid and c(bind) them
-    dnames <- dimnames(as.array(x))
-    ## NULL if scalar value, which means no "group_by"
+    # If there are dimnames, expand.grid and bind them. We also change the names
+    # of the two array dimensions to avoid duplicated variable names in  the
+    # tibble.
     if (!is.null(dnames)) {
-        dims <- do.call(expand.grid, dnames)
-        vals <- c(dims, vals)
+        types <- getDimTypes(x)
+
+        # Change MR selection vars to T/F/NA
+        is_selected <- types == "mr_selections"
```

Comment:
FWIW, there is a `crunch` function `is.selectedDimension` which does exactly this. I'm not sure it's necessary to import and use it when you can accomplish it in one line, but it would be using the approved machinery from `crunch` so if that ever changes for some reason (I can't imagine what that would be!) it would be automatically updated.

Reply:
I don't think it's exported, so I'll just keep this for now.